draft-irtf-cfrg-hybrid-kems.md

```diff
@@ -537,175 +545,326 @@ and L arguments.
+incompatibilities in cases where a private key needs to be moved from one system
+to another.
 
-    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)
-    dk_T = Group_T.RandomScalar(seed_T)
-    ek_T = Group_T.Exp(Group_T.g, dk_T)
+The security impact of this deviation is to reduce binding properties from
+MAL-BIND-P-Q to LEAK-BIND-P-Q. As discussed below, binding properties can
+address a variety of attack scenarios, including LEAK scenarios in which an
+attacker has passive access to the decapsulation key and MAL scenarios in which
+an attacker can cause the victim to use a crafted decapsulation key. The above
+hybrid KEM framework assures binding properties in the face of a LEAK attacker,
+irrespective of how key generation is done. The additional provided by the
+default "shared seed" key generation upgrades this to protection against a MAL
+attacker.
 
-    return (ek_PQ, ek_T, dk_PQ, dk_T)
+As a result, this alternative style of key generation should only be used in
+environments where implementations of component algorithms do not allow
+decapsulation keys to be imported or exported.  When decapsulation keys are
+moved around, they MUST always use the seed format.
+
+## UG Framework: Universal Combiner with a Nominal Group
 
+This framework combines a PQ KEM with a nominal group, using the universal
+combiner function.  It should be used in cases where the application wants to
+use a nominal group for the traditional component, and does not want to rely on
+the C2PRI assumption for the PQ KEM.
+
+~~~
 def DeriveKeyPair(seed):
-    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsulationKey(seed)
+    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyG(seed)
     return (concat(ek_PQ, ek_T), seed)
 
-def GenerateKeyPair():
-    seed = random(Nseed)
-    return DeriveKeyPair(seed)
+def Encaps(ek):
+    (ek_PQ, ek_T) = split(KEM_PQ.Nek, Group_T.Nelem, ek)
+    (ss_PQ, ss_T, ct_PQ, ct_T) = prepareEncapG(ek_PQ, ek_T)
+    ss_H = UniversalCombiner(ss_PQ, ss_T, ct_T, ek_T, Label))
+    ct_H = concat(ct_PQ, ct_T)
+    return (ss_H, ct_H)
+
+def Decaps(dk, ct):
+    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyG(dk)
+    (ss_PQ, ss_T) = prepareDecapG(ct_PQ, ct_T, dk_PQ, dk_T)
+    ss_H = UniversalCombiner(ss_PQ, ss_T, ct_T, ek_T, Label))
+    return ss_H
+~~~
+
+## UK Framework: Universal Combiner with a KEM
+
+This framework combines a PQ KEM with a traditional KEM, using the universal
+combiner function.  It should be used in cases where the application wants to
+use a KEM for the traditional component, and does not want to rely on the C2PRI
+assumption for the PQ KEM.
+
+~~~
+def DeriveKeyPair(seed):
+    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyK(seed)
+    return (concat(ek_PQ, ek_T), seed)
 
 def Encaps(ek):
     (ek_PQ, ek_T) = split(KEM_PQ.Nek, Group_T.Nelem, ek)
+    (ss_PQ, ss_T, ct_PQ, ct_T) = prepareEncapK(ek_PQ, ek_T)
+    ss_H = UniversalCombiner(ss_PQ, ss_T, ct_T, ek_T, Label))
+    ct_H = concat(ct_PQ, ct_T)
```

Comment:
We do `concat()` on their own lines in some places but not others, should we try to be consistent?